execution/executor/src/lib.rs

```diff
@@ -282,38 +282,31 @@ where
             .map(|(idx, _)| idx + 1);
         let transaction_count = new_epoch_marker.unwrap_or(vm_outputs.len());
 
-        let txn_blobs =
-            itertools::zip_eq(vm_outputs.iter(), transactions.iter())
-                .take(transaction_count)
-                .map(|(vm_output, txn)| {
-                    process_write_set(
-                        txn,
-                        &mut account_to_state,
-                        vm_output.write_set().clone(),
-                    )
-                })
-                .collect::<Result<Vec<_>>>()?;
-
-        let mut current_state_tree = Arc::clone(parent_trees.state_tree());
-        let mut txn_state_trees = vec![];
-        for updated_blobs in txn_blobs.iter() {
-            current_state_tree = Arc::new(
-                current_state_tree
-                    .update(
-                        updated_blobs
-                            .iter()
-                            .map(|(addr, value)| (addr.hash(), value.clone()))
-                            .collect(),
-                        &proof_reader,
-                    )
-                    .expect("Failed to update state tree."),
-            );
-            txn_state_trees.push(current_state_tree.clone());
-        }
+        let txn_blobs = itertools::zip_eq(vm_outputs.iter(), transactions.iter())
+            .take(transaction_count)
+            .map(|(vm_output, txn)| {
+                process_write_set(txn, &mut account_to_state, vm_output.write_set().clone())
+            })
+            .collect::<Result<Vec<_>>>()?;
 
-        for ((vm_output, txn), (state_tree, blobs)) in itertools::zip_eq(
+        let (txn_state_roots, current_state_tree) = parent_trees
+            .state_tree()
+            .batch_update(
+                txn_blobs
+                    .iter()
+                    .map(|m| {
+                        m.iter()
+                            .map(|(account, value)| (account.hash(), value.clone()))
```

Comment:
As we have discussed earlier, I think for the new algorithm we should pass immutable reference to the blobs here, and it's going to be cloned all the way at the end when the leaves are created, because we wouldn't be able to move a single element out of the vector at a specified index (as we are passing a vector of all updates down and partitioning the indices).
Before this was fine because we were just moving all the individual blobs and calling update_one.

Your simple batch_update implementation would also have to clone internally.

Reply:
In fact my interface was flattened Vec<(Hashvalue, usize, &V)> where usize is txn_id, but we could use either one, let's decide later.

Reply:
Updated the api! Lemme know how the comment there sounds to you (if there needs more explanation why we prefer a reference in a batch api vs a owned instance)

Reply:
Thanks! Regarding the comment, I would say instead of:
    /// The `update_batch` will take in a reference of value instead of an owned instance. This is
    /// because it would be nicer for future parallelism.	    
    /// because it would be nicer for future parallelism.
Maybe something like:
    /// The `update_batch` will take in a reference of value instead of an owned instance. This is
    /// to avoid an extra clone of each value, since an upcoming optimized implementation of this
    /// interface has to clone the value during the leaf creation.
But feel free to adjust as you like, thanks